// @inlang/sdk-js/dist/adapter-sveltekit/runtime/client/reactive/context.js
import { setContext } from "svelte";
import { writable } from "svelte/store";
import { inlangSymbol } from "../../shared/utils.js";
import { getRuntimeFromContext as getRuntimeFromContextShared } from "../shared/context.js";
export const getRuntimeFromContext = () => getRuntimeFromContextShared();
export const addRuntimeToContext = (runtime) => {
    const _language = writable(runtime.language);
    const _i = writable(runtime.i);
    const switchLanguage = async (language) => {
        if (runtime.language === language)
            return;
        await runtime.loadResource(language);
        runtime.switchLanguage(language);
        _i.set(runtime.i);
        _language.set(language);
    };
    setContext(inlangSymbol, {
        language: readonly(_language),
        referenceLanguage: runtime.referenceLanguage,
        languages: runtime.languages,
        i: readonly(_i),
        loadResource: runtime.loadResource,
        switchLanguage,
        route,
    });
};
// TODO: output warning that calling this does not make sense
const route = (href) => href;
// ------------------------------------------------------------------------------------------------
// copy from "svelte/store" to support older versions than `3.56.0`
function readonly(store) {
    return {
        // @ts-ignore
        subscribe: store.subscribe.bind(store),
    };
}
